dali/core/fast_div_test.cu

```diff
@@ -0,0 +1,300 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/core/fast_div.h"  // NOLINT
+#include <gtest/gtest.h>
+#include <random>
+#include <iostream>
+#include "dali/test/device_test.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/cuda_event.h"
+
+namespace dali {
+
+template <uint32_t divisor>
+void TestDiv32(bool quick_test) {
+  fast_div<uint32_t> fast = divisor;
+
+  std::cerr << "Testing x / " << divisor << std::endl;
+  uint32_t range = quick_test ? (1<<20) : 0xFFFFFFFFu;
+  for (uint32_t xx = 0; xx < range; xx++) {
+    uint32_t x = quick_test
+      ? ((xx >> 12) << 24) | (xx & 0xfff)
+      : xx;
+    ASSERT_EQ(x / divisor, x / fast) << " when dividing " << x << " / " << divisor;
+  }
+
+  uint32_t x = 0xFFFFFFFEu;
+  ASSERT_EQ(x / divisor, x / fast) << " when dividing " << x << " / " << divisor;
+}
+
+template <uint64_t divisor>
+void TestDiv64(bool quick_test) {
+  fast_div<uint64_t> fast = divisor;
+  std::cerr << "Testing x / " << divisor << std::endl;
+  uint64_t range = quick_test ? (1<<20) : 0xFFFFFFFFu;
+  for (uint64_t xx = 0; xx < range; xx++) {
+    uint64_t x = quick_test
+      ? ((xx >> 12) << 56) + (xx & 0xfff)
+      : ((xx >> 12) << 44) + (xx & 0xfff);
+    ASSERT_EQ(x / divisor, x / fast) << " when dividing " << x << " / " << divisor;
+  }
+  uint64_t x = 0xFFFFFFFFFFFFFFFEuL;
+  ASSERT_EQ(x / divisor, x / fast) << " when dividing " << x << " / " << divisor;
+}
+
+void TestDiv32(bool quick) {
+  // use compile-time constants so the compiler can optimize the reference division
+  TestDiv32<0xFFFFFFFEu>(quick);
+  TestDiv32<3>(quick);
+  TestDiv32<5>(quick);
+  TestDiv32<7>(quick);
+  TestDiv32<1>(quick);
+  TestDiv32<0x800000>(quick);
+  TestDiv32<19>(quick);
+  TestDiv32<42>(quick);
+  TestDiv32<12345678>(quick);
+  TestDiv32<0x82345678>(quick);
+}
+
+void TestDiv64(bool quick) {
+  // use compile-time constants so the compiler can optimize the reference division
+  TestDiv64<0xFFFFFFFFFFFFFFFEuL>(quick);
+  TestDiv64<3>(quick);
+  TestDiv64<5>(quick);
+  TestDiv64<7>(quick);
+  TestDiv64<1>(quick);
+  TestDiv64<0x800000>(quick);
+  TestDiv64<0x80000000000000uL>(quick);
+  TestDiv64<19>(quick);
+  TestDiv64<42>(quick);
+  TestDiv64<12345678901234567uL>(quick);
+  TestDiv64<0x82345678DEADBEEFuL>(quick);
+}
+
+// This test is disabled because it's ridiculously slow - only run when touching fast_div
+TEST(FastDiv, DISABLED_U32_Host_Slow) {
+  TestDiv32(false);
+}
```

Comment:
This is here just in case someone decides to touch this code (and see the world burn) - if something totally breaks apart, it's very likely that the faster test will catch it.